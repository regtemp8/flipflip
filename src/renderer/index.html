<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>FlipFlip</title>
    <!-- Responsive meta tag -- ensures proper rendering and touch zooming for all devices -->
    <meta name="viewport" content="minimum-scale=1, initial-scale=1, width=device-width, shrink-to-fit=no" />
    <!-- Roboto Font -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap" />
    <!-- Font Icons -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons" />

    <meta
      http-equiv="Content-Security-Policy"
      content="frame-src https://www.imagefap.com http://www.imagefap.com https://hypno.nimja.com/ http://hypno.nimja.com/; font-src 'self' https://fonts.gstatic.com; style-src 'self' https://fonts.googleapis.com 'unsafe-inline'"
    />
  </head>

  <body>
    <div id="app" class="app"></div>
  </body>
</html>
