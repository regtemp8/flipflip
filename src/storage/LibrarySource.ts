import type Tag from './Tag'
import type Clip from './Clip'
import { copy } from '../renderer/data/utils'

export default interface LibrarySource {
  id: number
  url?: string
  offline: boolean
  marked: boolean
  lastCheck: string
  tags: Tag[]
  clips: Clip[]
  disabledClips: number[]
  blacklist: string[]
  count: number
  countComplete: boolean
  weight: number

  // Type specific properties
  // Local
  dirOfSources: boolean
  // Video
  subtitleFile?: string
  duration?: number
  resolution?: number
  // Reddit
  redditFunc?: string
  redditTime?: string
  // Twitter
  includeRetweets: boolean
  includeReplies: boolean
}

const initialLibrarySource: LibrarySource = {
  id: 0,
  offline: false,
  marked: false,
  lastCheck: null,
  tags: [],
  clips: [],
  disabledClips: [],
  blacklist: [],
  count: 0,
  countComplete: false,
  weight: 1,
  dirOfSources: false,
  includeRetweets: false,
  includeReplies: false
}

export function newLibrarySource (init?: Partial<LibrarySource>) {
  return Object.assign(copy<LibrarySource>(initialLibrarySource), init)
}
