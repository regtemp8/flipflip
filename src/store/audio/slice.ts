import { createSlice, type PayloadAction } from '@reduxjs/toolkit'
import type Audio from './Audio'
import { getAudio } from './selectors'
import {
  type EntryState,
  type EntryUpdate,
  setEntrySlice,
  setEntry,
  updateEntry
} from '../EntryState'

export const initialAudioState: EntryState<Audio> = {
  name: 'audioSlice',
  nextID: 1,
  entries: {}
}
export const audioSlice = createSlice({
  name: 'audios',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState: initialAudioState,
  reducers: {
    setAudioSlice: (state, action: PayloadAction<EntryState<Audio>>) => {
      setEntrySlice(state, action.payload)
    },
    setAudio: (state, action: PayloadAction<Audio>) => {
      setEntry(state, action.payload)
    },
    updateAudio: (state, action: PayloadAction<Audio>) => {
      const id = action.payload.id
      Object.assign(state.entries[id], action.payload)
    },
    setAudioTickTF: (state, action: PayloadAction<EntryUpdate<string>>) => {
      getAudio(state, action.payload.id).tickTF = action.payload.value
    },
    setAudioTickDuration: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      getAudio(state, action.payload.id).tickDelay = action.payload.value
    },
    setAudioTickDurationMin: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      getAudio(state, action.payload.id).tickMinDelay = action.payload.value
    },
    setAudioTickDurationMax: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      getAudio(state, action.payload.id).tickMaxDelay = action.payload.value
    },
    setAudioTickSinRate: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      getAudio(state, action.payload.id).tickSinRate = action.payload.value
    },
    setAudioTickBPMMulti: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      getAudio(state, action.payload.id).tickBPMMulti = action.payload.value
    },
    setAudioStopAtEnd: (state, action: PayloadAction<EntryUpdate<boolean>>) => {
      getAudio(state, action.payload.id).stopAtEnd = action.payload.value
    },
    setAudioNextSceneAtEnd: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getAudio(state, action.payload.id).nextSceneAtEnd = action.payload.value
    },
    setAudioTick: (state, action: PayloadAction<EntryUpdate<boolean>>) => {
      getAudio(state, action.payload.id).tick = action.payload.value
    },
    setAudioSpeed: (state, action: PayloadAction<EntryUpdate<number>>) => {
      getAudio(state, action.payload.id).speed = action.payload.value
    },
    setAudioVolume: (state, action: PayloadAction<EntryUpdate<number>>) => {
      getAudio(state, action.payload.id).volume = action.payload.value
    },
    setAudioUrl: (state, action: PayloadAction<EntryUpdate<string>>) => {
      getAudio(state, action.payload.id).url = action.payload.value
    },
    setAudioTags: (state, action: PayloadAction<EntryUpdate<number[]>>) => {
      getAudio(state, action.payload.id).tags = action.payload.value
    },
    setAudioBPM: (state, action: PayloadAction<EntryUpdate<number>>) => {
      getAudio(state, action.payload.id).bpm = action.payload.value
    },
    setAudioToggleTag: (state, action: PayloadAction<EntryUpdate<number>>) => {
      const audio = getAudio(state, action.payload.id)
      const index = audio.tags.indexOf(action.payload.value)
      if (index === -1) {
        audio.tags.push(action.payload.value)
      } else {
        audio.tags.splice(index, 1)
      }
    },
    setAudioMarked: (state, action: PayloadAction<EntryUpdate<boolean>>) => {
      getAudio(state, action.payload.id).marked = action.payload.value
    },
    playTrack: (state, action: PayloadAction<number>) => {
      state.entries[action.payload].playedCount++
    }
  }
})

export const {
  setAudioSlice,
  setAudio,
  updateAudio,
  setAudioTickTF,
  setAudioTickDuration,
  setAudioTickDurationMin,
  setAudioTickDurationMax,
  setAudioTickSinRate,
  setAudioTickBPMMulti,
  setAudioStopAtEnd,
  setAudioNextSceneAtEnd,
  setAudioTick,
  setAudioSpeed,
  setAudioVolume,
  setAudioUrl,
  setAudioTags,
  setAudioBPM,
  setAudioToggleTag,
  setAudioMarked,
  playTrack
} = audioSlice.actions

export default audioSlice.reducer
