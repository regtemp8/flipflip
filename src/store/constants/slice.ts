import { createSlice, type PayloadAction } from '@reduxjs/toolkit'
import type SystemConstants from './SystemConstants'
import { initialSystemConstants } from './SystemConstants'

export const constantsSlice = createSlice({
  name: 'constants',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState: initialSystemConstants,
  reducers: {
    setConstants: (state, action: PayloadAction<SystemConstants>) => {
      Object.assign(state, action.payload)
    }
  }
})

export const { setConstants } = constantsSlice.actions

export default constantsSlice.reducer
