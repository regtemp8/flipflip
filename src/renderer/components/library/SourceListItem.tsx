import React, { useEffect, useState } from 'react'
import clsx from 'clsx'

import {
  Badge,
  Checkbox,
  Chip,
  Fab,
  IconButton,
  ListItem,
  ListItemAvatar,
  ListItemSecondaryAction,
  ListItemText,
  SvgIcon,
  TextField,
  type Theme,
  Tooltip,
  Typography
} from '@mui/material'

import createStyles from '@mui/styles/createStyles'
import withStyles, { type WithStyles } from '@mui/styles/withStyles'

import BuildIcon from '@mui/icons-material/Build'
import DeleteIcon from '@mui/icons-material/Delete'
import OfflineBoltIcon from '@mui/icons-material/OfflineBolt'

import { getCachePath, getTimestamp, urlToPath } from '../../data/utils'
import { getFileName, getSourceType } from '../player/Scrapers'
import { SDT, ST } from '../../data/const'
import SourceIcon from './SourceIcon'
import { grey } from '@mui/material/colors'
import { useAppDispatch, useAppSelector } from '../../../store/hooks'
import { systemMessage } from '../../../store/app/slice'
import { selectConstants } from '../../../store/constants/selectors'
import {
  selectAppConfigCachingDirectory,
  selectAppConfigCachingEnabled,
  selectAppTutorial
} from '../../../store/app/selectors'
import {
  blacklistFile,
  clipVideo,
  downloadSource,
  playSceneFromLibrary
} from '../../../store/app/thunks'
import TagChip from './TagChip'
import {
  selectLibrarySourceURL,
  selectLibrarySourceWeight,
  selectLibrarySourceClips,
  selectLibrarySourceOffline,
  selectLibrarySourceMarked,
  selectLibrarySourceDuration,
  selectLibrarySourceResolution,
  selectLibrarySourceCount,
  selectLibrarySourceCountComplete,
  selectLibrarySourceTags,
  selectLibrarySourceDisabledClips,
  selectLibrarySourceBlacklist
} from '../../../store/librarySource/selectors'

const styles = (theme: Theme) =>
  createStyles({
    root: {
      display: 'flex'
    },
    oddChild: {
      backgroundColor:
        theme.palette.mode === 'light'
          ? (theme.palette.primary as any)['100']
          : grey[900],
      '&:hover': {
        backgroundColor:
          theme.palette.mode === 'light'
            ? (theme.palette.primary as any)['200']
            : '#080808'
      }
    },
    evenChild: {
      backgroundColor:
        theme.palette.mode === 'light'
          ? (theme.palette.primary as any)['50']
          : theme.palette.background.default,
      '&:hover': {
        backgroundColor:
          theme.palette.mode === 'light'
            ? (theme.palette.primary as any)['200']
            : '#080808'
      }
    },
    avatar: {
      backgroundColor: theme.palette.primary.main,
      boxShadow: 'none'
    },
    markedSource: {
      backgroundColor: theme.palette.secondary.main
    },
    sourceIcon: {
      color: theme.palette.primary.contrastText
    },
    sourceMarkedIcon: {
      color: theme.palette.secondary.contrastText
    },
    deleteButton: {
      backgroundColor: theme.palette.error.main
    },
    deleteIcon: {
      color: theme.palette.error.contrastText
    },
    errorIcon: {
      color: theme.palette.error.main,
      backgroundColor: theme.palette.error.contrastText,
      borderRadius: '50%'
    },
    importBadge: {
      zIndex: theme.zIndex.fab + 1
    },
    actionButton: {
      marginLeft: theme.spacing(1)
    },
    countChip: {
      userSelect: 'none',
      marginRight: theme.spacing(1),
      [theme.breakpoints.down('sm')]: {
        display: 'none'
      }
    },
    fullTag: {
      [theme.breakpoints.down('md')]: {
        display: 'none'
      }
    },
    simpleTag: {
      [theme.breakpoints.up('md')]: {
        display: 'none'
      },
      [theme.breakpoints.down('sm')]: {
        display: 'none'
      }
    },
    urlField: {
      width: '100%',
      margin: 0
    },
    highlight: {
      borderWidth: 2,
      borderColor: theme.palette.secondary.main,
      borderStyle: 'solid'
    },
    disable: {
      pointerEvents: 'none'
    },
    noUserSelect: {
      userSelect: 'none'
    }
  })

export interface SourceListItemProps extends WithStyles<typeof styles> {
  checked: boolean
  index: number
  isEditing: number
  isLibrary: boolean
  isSelect: boolean
  source: number
  sources: number[]
  style: any
  useWeights?: boolean
  onClean: (sourceURL: string) => void
  onDelete: (sourceID: number) => void
  onEditBlacklist: (sourceID: number) => void
  onEndEdit: (newURL: string) => void
  onOpenClipMenu: (sourceID: number) => void
  onOpenWeightMenu: (sourceID: number) => void
  onRemove: (sourceID: number) => void
  onSourceOptions: (sourceID: number) => void
  onStartEdit: (sourceID: number) => void
  onToggleSelect: () => void
  savePosition: () => void
}

function SourceListItem(props: SourceListItemProps) {
  const dispatch = useAppDispatch()
  const { isWin32, pathSep } = useAppSelector(selectConstants())
  const tutorial = useAppSelector(selectAppTutorial())
  const cachingEnabled = useAppSelector(selectAppConfigCachingEnabled())
  const cachingDirectory = useAppSelector(selectAppConfigCachingDirectory())
  const url = useAppSelector(selectLibrarySourceURL(props.source))
  const offline = useAppSelector(selectLibrarySourceOffline(props.source))
  const marked = useAppSelector(selectLibrarySourceMarked(props.source))
  const weight = useAppSelector(selectLibrarySourceWeight(props.source))
  const duration = useAppSelector(selectLibrarySourceDuration(props.source))
  const resolution = useAppSelector(selectLibrarySourceResolution(props.source))
  const count = useAppSelector(selectLibrarySourceCount(props.source))
  const countComplete = useAppSelector(
    selectLibrarySourceCountComplete(props.source)
  )
  const tags = useAppSelector(selectLibrarySourceTags(props.source))
  const clips = useAppSelector(selectLibrarySourceClips(props.source))
  const disabledClips = useAppSelector(
    selectLibrarySourceDisabledClips(props.source)
  )
  const blacklist = useAppSelector(selectLibrarySourceBlacklist(props.source))

  const [urlInput, setUrlInput] = useState<string>()

  useEffect(() => {
    setUrlInput(url)
  }, [props.source])

  const onSourceIconClick = async (e: MouseEvent) => {
    const sourceURL = url as string
    const sourceType = getSourceType(sourceURL)
    if (e.shiftKey && e.ctrlKey && e.altKey) {
      if (
        sourceType === ST.local ||
        sourceType === ST.video ||
        sourceType === ST.playlist ||
        sourceType === ST.list
      ) {
        const exists = await window.flipflip.api.pathExists(sourceURL)
        if (exists) {
          props.onDelete(props.source)
        }
      }
    } else if (e.shiftKey && !e.ctrlKey && !e.altKey) {
      openExternalURL(sourceURL)
    } else if (!e.shiftKey && !e.ctrlKey && e.altKey) {
      // If local source, still catch keypress, but don't do anything
      if (
        sourceType !== ST.local &&
        sourceType !== ST.video &&
        sourceType !== ST.piwigo &&
        sourceType !== ST.hydrus &&
        sourceType !== ST.nimja
      ) {
        dispatch(downloadSource(props.source))
      }
    } else if (!e.shiftKey && e.ctrlKey && !e.altKey) {
      const fileType = getSourceType(sourceURL)
      let cachePath = await getCachePath(cachingDirectory, sourceURL)
      if (fileType === ST.video || fileType === ST.playlist) {
        if (
          !(await window.flipflip.api.pathExists(
            cachePath + getFileName(sourceURL, pathSep)
          ))
        ) {
          cachePath = undefined
          window.flipflip.api.showItemInFolder(sourceURL)
        }
      }
      if (cachePath) {
        openDirectory(cachePath)
      }
    } else if (!e.shiftKey && !e.ctrlKey) {
      props.savePosition()
      try {
        dispatch(playSceneFromLibrary(props.source, props.sources))
      } catch (e) {
        dispatch(systemMessage('The source ' + url + " isn't in your Library"))
      }
    }
  }

  const onClip = () => {
    props.savePosition()
    dispatch(clipVideo(props.source, props.sources))
  }

  const onClickBlacklist = (e: MouseEvent) => {
    if (e.shiftKey) {
      clearBlacklist(url as string)
    } else {
      props.onEditBlacklist(props.source)
    }
  }

  const clearBlacklist = (sourceURL: string) => {
    dispatch(blacklistFile(sourceURL, undefined))
  }

  const onStartEdit = (sourceID: number) => {
    props.onStartEdit(sourceID)
  }

  const onEditSource = (e: MouseEvent) => {
    const input = e.target as HTMLInputElement
    setUrlInput(input.value)
  }

  const onEndEdit = () => {
    props.onEndEdit(urlInput as string)
  }

  const openDirectory = (cachePath: string) => {
    if (isWin32) {
      openExternalURL(cachePath)
    } else {
      openExternalURL(urlToPath(cachePath, isWin32))
    }
  }

  const openExternalURL = (url: string) => {
    window.flipflip.api.openExternal(url)
  }

  const classes = props.classes
  const sourceType = url && getSourceType(url)
  return (
    <div
      style={props.style}
      className={clsx(
        props.index % 2 === 0 ? classes.evenChild : classes.oddChild,
        tutorial === SDT.source && classes.highlight,
        tutorial && classes.disable
      )}
    >
      <ListItem>
        {props.isSelect && (
          <Checkbox
            value={props.source}
            onChange={props.onToggleSelect.bind(this)}
            checked={props.checked}
          />
        )}
        <ListItemAvatar>
          <Badge
            classes={{
              badge: classes.importBadge
            }}
            invisible={!offline}
            overlap="circular"
            anchorOrigin={{
              vertical: 'top',
              horizontal: 'left'
            }}
            badgeContent={<OfflineBoltIcon className={classes.errorIcon} />}
          >
            <Tooltip
              disableInteractive
              title={
                <div>
                  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Click:
                  Library Tagging
                  <br />
                  Shift+Click: Open Source
                  <br />
                  &nbsp;&nbsp;Ctrl+Click:{' '}
                  {sourceType === ST.video ? 'Reveal File' : 'Open Cache'}
                  {sourceType !== ST.local &&
                    sourceType !== ST.video &&
                    sourceType !== ST.piwigo &&
                    sourceType !== ST.hydrus &&
                    sourceType !== ST.nimja && (
                      <React.Fragment>
                        <br />
                        &nbsp;&nbsp;&nbsp;Alt+Click: Download Source
                      </React.Fragment>
                    )}
                </div>
              }
            >
              <Fab
                size="small"
                onClick={onSourceIconClick.bind(this)}
                className={clsx(
                  classes.avatar,
                  marked && classes.markedSource,
                  tutorial === SDT.sourceAvatar && classes.highlight
                )}
              >
                <SourceIcon
                  url={url}
                  className={clsx(
                    classes.sourceIcon,
                    marked && classes.sourceMarkedIcon
                  )}
                />
              </Fab>
            </Tooltip>
          </Badge>
        </ListItemAvatar>

        <ListItemText classes={{ primary: classes.root }}>
          {props.isEditing === props.source && (
            <form onSubmit={onEndEdit.bind(this)} className={classes.urlField}>
              <TextField
                variant="standard"
                autoFocus
                fullWidth
                value={urlInput}
                margin="none"
                className={classes.urlField}
                onBlur={onEndEdit.bind(this)}
                onChange={onEditSource.bind(this)}
              />
            </form>
          )}
          {props.isEditing !== props.source && (
            <React.Fragment>
              <Typography
                noWrap
                className={clsx(
                  classes.noUserSelect,
                  tutorial === SDT.sourceTitle && classes.highlight
                )}
                onClick={onStartEdit.bind(this, props.source)}
              >
                {url}
              </Typography>
              {tags &&
                tags.map((tagID) => (
                  <React.Fragment key={tagID}>
                    <TagChip
                      tagID={tagID}
                      className={clsx(
                        classes.noUserSelect,
                        classes.actionButton,
                        classes.fullTag,
                        tutorial === SDT.sourceTags && classes.highlight
                      )}
                      outlined
                    />
                    <TagChip
                      tagID={tagID}
                      className={clsx(
                        classes.noUserSelect,
                        classes.actionButton,
                        classes.simpleTag,
                        tutorial === SDT.sourceTags && classes.highlight
                      )}
                      outlined
                      simpleTag
                    />
                  </React.Fragment>
                ))}
            </React.Fragment>
          )}
        </ListItemText>

        {props.isEditing !== props.source && (
          <ListItemSecondaryAction
            className={clsx(
              tutorial === SDT.sourceButtons && classes.highlight
            )}
          >
            {props.useWeights && (
              <React.Fragment>
                <Chip
                  className={classes.countChip}
                  clickable
                  label={weight || '1'}
                  color="default"
                  size="small"
                  onClick={props.onOpenWeightMenu.bind(this, props.source)}
                />
              </React.Fragment>
            )}
            {sourceType === ST.video && duration && (
              <Chip
                className={classes.countChip}
                label={getTimestamp(duration)}
                color="secondary"
                size="small"
              />
            )}
            {sourceType === ST.video && !!resolution && (
              <Chip
                className={classes.countChip}
                label={`${resolution}p`}
                color="secondary"
                size="small"
              />
            )}
            {count > 0 && sourceType !== ST.video && (
              <Chip
                className={clsx(
                  classes.countChip,
                  tutorial === SDT.sourceCount && classes.highlight
                )}
                label={`${count}${countComplete ? '' : '+'}`}
                color="primary"
                size="small"
              />
            )}
            {!props.isLibrary &&
              clips &&
              clips.length > 0 &&
              sourceType === ST.video && (
                <Chip
                  className={classes.countChip}
                  label={
                    (disabledClips
                      ? clips.filter(
                          (clipID) => !disabledClips.includes(clipID)
                        )
                      : clips
                    ).length +
                    '/' +
                    clips.length
                  }
                  onClick={props.onOpenClipMenu.bind(this, props.source)}
                  color="primary"
                  size="small"
                />
              )}
            {props.isLibrary &&
              clips &&
              clips.length > 0 &&
              sourceType === ST.video && (
                <Chip
                  className={classes.countChip}
                  label={clips.length}
                  color="primary"
                  size="small"
                />
              )}
            {(sourceType === ST.local ||
              sourceType === ST.video ||
              sourceType === ST.twitter ||
              sourceType === ST.reddit) && (
              <IconButton
                onClick={props.onSourceOptions.bind(this, props.source)}
                className={classes.actionButton}
                edge="end"
                size="small"
                aria-label="options"
              >
                <BuildIcon />
              </IconButton>
            )}
            {sourceType === ST.video && (
              <IconButton
                onClick={onClip.bind(this)}
                className={classes.actionButton}
                edge="end"
                size="small"
                aria-label="clip"
              >
                <SvgIcon>
                  <path
                    d="M11 21H7V19H11V21M15.5 19H17V21H13V19H13.2L11.8 12.9L9.3 13.5C9.2 14 9 14.4 8.8
                        14.8C7.9 16.3 6 16.7 4.5 15.8C3 14.9 2.6 13 3.5 11.5C4.4 10 6.3 9.6 7.8 10.5C8.2 10.7 8.5
                        11.1 8.7 11.4L11.2 10.8L10.6 8.3C10.2 8.2 9.8 8 9.4 7.8C8 6.9 7.5 5 8.4 3.5C9.3 2 11.2
                        1.6 12.7 2.5C14.2 3.4 14.6 5.3 13.7 6.8C13.5 7.2 13.1 7.5 12.8 7.7L15.5 19M7 11.8C6.3
                        11.3 5.3 11.6 4.8 12.3C4.3 13 4.6 14 5.3 14.4C6 14.9 7 14.7 7.5 13.9C7.9 13.2 7.7 12.2 7
                        11.8M12.4 6C12.9 5.3 12.6 4.3 11.9 3.8C11.2 3.3 10.2 3.6 9.7 4.3C9.3 5 9.5 6 10.3 6.5C11
                        6.9 12 6.7 12.4 6M12.8 11.3C12.6 11.2 12.4 11.2 12.3 11.4C12.2 11.6 12.2 11.8 12.4
                        11.9C12.6 12 12.8 12 12.9 11.8C13.1 11.6 13 11.4 12.8 11.3M21 8.5L14.5 10L15 12.2L22.5
                        10.4L23 9.7L21 8.5M23 19H19V21H23V19M5 19H1V21H5V19Z"
                  />
                </SvgIcon>
              </IconButton>
            )}
            {blacklist && blacklist.length > 0 && (
              <IconButton
                onClick={onClickBlacklist.bind(this)}
                className={classes.actionButton}
                edge="end"
                size="small"
                aria-label="clear blacklist"
              >
                <SvgIcon>
                  <path
                    d="M2 6V8H14V6H2M2 10V12H11V10H2M14.17 10.76L12.76 12.17L15.59 15L12.76 17.83L14.17
                        19.24L17 16.41L19.83 19.24L21.24 17.83L18.41 15L21.24 12.17L19.83 10.76L17 13.59L14.17
                        10.76M2 14V16H11V14H2Z"
                  />
                </SvgIcon>
              </IconButton>
            )}
            {cachingEnabled &&
              sourceType !== ST.local &&
              ((sourceType !== ST.video && sourceType !== ST.playlist) ||
                /^https?:\/\//g.exec(url as string) != null) && (
                <React.Fragment>
                  <IconButton
                    onClick={props.onClean.bind(this, url)}
                    className={classes.actionButton}
                    edge="end"
                    size="small"
                    aria-label="clean cache"
                  >
                    <SvgIcon>
                      <path
                        d="M19.36 2.72L20.78 4.14L15.06 9.85C16.13 11.39 16.28 13.24 15.38 14.44L9.06
                          8.12C10.26 7.22 12.11 7.37 13.65 8.44L19.36 2.72M5.93 17.57C3.92 15.56 2.69 13.16 2.35
                          10.92L7.23 8.83L14.67 16.27L12.58 21.15C10.34 20.81 7.94 19.58 5.93 17.57Z"
                      />
                    </SvgIcon>
                  </IconButton>
                </React.Fragment>
              )}
            <IconButton
              onClick={props.onRemove.bind(this, props.source)}
              className={clsx(classes.deleteButton, classes.actionButton)}
              edge="end"
              size="small"
              aria-label="delete"
            >
              <DeleteIcon className={classes.deleteIcon} color="inherit" />
            </IconButton>
          </ListItemSecondaryAction>
        )}
      </ListItem>
    </div>
  )
}

;(SourceListItem as any).displayName = 'SourceListItem'
export default withStyles(styles)(SourceListItem as any)
