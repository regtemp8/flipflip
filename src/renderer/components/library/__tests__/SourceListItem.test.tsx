import React from "react";
import { describe, it, expect } from "@jest/globals";
import renderer from "react-test-renderer";
import SourceListItem from "../SourceListItem";
import TestProvider from "../../../../../test/util/TestProvider";
import store from "../../../../store/store";
import { setLibrarySource } from "../../../../store/librarySource/slice";
import { newLibrarySource } from "../../../../store/librarySource/LibrarySource";

jest.mock('../SourceIcon', () => 'SourceIcon');

// TODO create functional tests instead of snapshots
describe("SourceListItem", () => {
  it("should match snapshot", () => {
    const sourceID = 3
    store.dispatch(setLibrarySource(newLibrarySource({id: sourceID, url: 'image.png'})))
    const component = renderer.create(
      <TestProvider store={store}>
        <SourceListItem
          checked={false}
          index={0}
          isEditing={null}
          isLibrary={false}
          isSelect={false}
          source={sourceID}
          sources={[]}
          style={null}
          onClean={(source) => {}}
          onDelete={(source) => {}}
          onEditBlacklist={(source) => {}}
          onEndEdit={(newURL) => {}}
          onOpenClipMenu={(source) => {}}
          onOpenWeightMenu={(source) => {}}
          onRemove={(source) => {}}
          onSourceOptions={(source) => {}}
          onStartEdit={(id) => {}}
          onToggleSelect={() => {}}
          savePosition={() => {}}
        />
      </TestProvider>
    );

    let tree = component.toJSON();
    // expect(tree).toMatchSnapshot();
  });
});
