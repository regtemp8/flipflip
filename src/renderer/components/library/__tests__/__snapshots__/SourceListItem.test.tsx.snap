// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceListItem should match snapshot 1`] = `
<div
  className="SourceListItem-evenChild-3"
  style={null}
>
  <li
    className="MuiListItem-container css-oyakqy-MuiListItem-container"
  >
    <div
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-secondaryAction css-nzmfr-MuiListItem-root"
      disabled={false}
    >
      <div
        className="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
      >
        <span
          className="MuiBadge-root css-1c32n2y-MuiBadge-root"
        >
          <button
            aria-label={null}
            aria-labelledby={null}
            className="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-default MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-default SourceListItem-avatar-4 css-12m9wit-MuiButtonBase-root-MuiFab-root"
            data-mui-internal-clone-element={true}
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <SourceIcon
              className="SourceListItem-sourceIcon-6"
              url="image.png"
            />
            <span
              className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <span
            className="MuiBadge-badge SourceListItem-importBadge-11 MuiBadge-standard MuiBadge-invisible MuiBadge-anchorOriginTopLeft MuiBadge-anchorOriginTopLeftCircular MuiBadge-overlapCircular css-1cyh125-MuiBadge-badge"
          />
        </span>
      </div>
      <div
        className="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary SourceListItem-root-1 css-10hburv-MuiTypography-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap SourceListItem-noUserSelect-19 css-fvtpii-MuiTypography-root"
            onClick={[Function]}
          >
            image.png
          </p>
        </span>
      </div>
    </div>
    <div
      className="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
    >
      <button
        aria-label="options"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeSmall SourceListItem-actionButton-12 css-1bkxqye-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="BuildIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m22.7 19-9.1-9.1c.9-2.3.4-5-1.5-6.9-2-2-5-2.4-7.4-1.3L9 6 6 9 1.6 4.7C.4 7.1.9 10.1 2.9 12.1c1.9 1.9 4.6 2.4 6.9 1.5l9.1 9.1c.4.4 1 .4 1.4 0l2.3-2.3c.5-.4.5-1.1.1-1.4z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="delete"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeSmall SourceListItem-deleteButton-8 SourceListItem-actionButton-12 css-1bkxqye-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium SourceListItem-deleteIcon-9 css-i4bv87-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </li>
</div>
`;
